browser-sign-in/app/src/main/java/com/okta/android/samples/browser_sign_in/data/UserDashboardViewModel.kt

```diff
@@ -0,0 +1,161 @@
+    private val _loginState = MutableLiveData<UserDashboardLoginState>(UserDashboardLoginState.Loading)
+    private val _userGreetingState = MutableLiveData<UserGreetingState>(UserGreetingState.Loading)
+    private val _userInfoState = MutableLiveData<UserInfoState>(UserInfoState.Empty)
+
+    val loginState: LiveData<UserDashboardLoginState> = _loginState
+    val userGreetingState: LiveData<UserGreetingState> = _userGreetingState
+    val userInfoState: LiveData<UserInfoState> = _userInfoState
+
+    init {
+        viewModelScope.launch {
+            refreshCurrentLoginState()
+        }
+    }
+
+    fun logoutOfBrowser(context: Context) {
+        viewModelScope.launch {
+            _loginState.value = UserDashboardLoginState.Loading
+
+            val result =
+                CredentialBootstrap.oidcClient.createWebAuthenticationClient().logoutOfBrowser(
+                    context = context,
+                    redirectUrl = BuildConfig.SIGN_OUT_REDIRECT_URI,
+                    CredentialBootstrap.defaultCredential().token?.idToken ?: "",
+                )
+            when (result) {
+                is OidcClientResult.Error -> {
+                    Timber.e(result.exception, "Failed to logout.")
+                    _loginState.value = UserDashboardLoginState.LogoutFailed(
+                        errorMessage = "Failed to logout.",
+                    )
+                }
+                is OidcClientResult.Success -> {
+                    CredentialBootstrap.defaultCredential().delete()
+                    _loginState.value = UserDashboardLoginState.LoggedOut
+                }
+            }
+        }
+    }
+
+    fun fetchUserInfo() {
+        viewModelScope.launch {
+            _userInfoState.value = UserInfoState.Loading
+            val credential = CredentialBootstrap.defaultCredential()
+
+            when (val result = credential.getUserInfo()) {
+                is OidcClientResult.Error -> {
+                    Timber.e(result.exception, "Failed to fetch userinfo.")
+                    _userInfoState.value = UserInfoState.Loaded(
+                        userInfoText = "Failed to fetch userinfo.",
+                    )
+                }
+                is OidcClientResult.Success -> {
+                    val userClaimsList =
+                        result.result.deserializeClaims(JsonObject.serializer()).asList()
+                    _userInfoState.value = UserInfoState.Loaded(
+                        userInfoText = userClaimsList.joinToString("\n"),
+                    )
+                }
+            }
+        }
+    }
+
+    private suspend fun refreshCurrentLoginState() {
+        val credential = CredentialBootstrap.defaultCredential()
+        if (credential.token == null) {
+            _userGreetingState.value = UserGreetingState.Loading
+            _loginState.value = UserDashboardLoginState.LoggedOut
+        } else {
+            val idToken = credential.idToken()!!
```

Comment:
I think my last comment wasn't quite clear here. I wasn't suggesting we assume the id token is non null, I was just suggesting we move it up the logic chain, and rather than saying `if (credential.token == null) {` on line 100, we'd do `if (idToken == null) {` instead. 

This would treat a user without an ID token as not logged in. Which is what I was targeting.

Reply:
Oh I see. That makes sense